pkg/setting/setting.go

```diff
@@ -450,15 +450,21 @@ type CommandLineArgs struct {
 func (cfg Cfg) parseAppUrlAndSubUrl(section *ini.Section) (string, string, error) {
 	appUrl := valueAsString(section, "root_url", "http://localhost:3000/")
 
-	if appUrl[len(appUrl)-1] != '/' {
-		appUrl += "/"
-	}
-
 	// Check if has app suburl.
 	url, err := url.Parse(appUrl)
 	if err != nil {
 		cfg.Logger.Error("Invalid root_url.", "url", appUrl, "error", err)
-		os.Exit(1)
+		return "", "", errors.New(fmt.Sprint("Invalid root_url:", err))
+	}
+
+	if len(url.Host) == 0 {
+		cfg.Logger.Error("Invalid root_url. root_url must have a host component.", "url", appUrl)
+		return "", "", errors.New(fmt.Sprint("Invalid root_url: ", appUrl, " root_url must have a host component."))
+	}
+
+	if len(url.Scheme) == 0 {
```

Comment:
How confident are you that adding these do not end up breaking any setup?

Reply:
After some more digging, not very. appUrl is used when redirecting or when constructing urls in alert notifications. If users don't encounter either of those things, their config could be any string without causing them visible problems. Definitely worth taking a step back. I'm going to close this because there are too many unknowns.